web-common/src/Halogen/Extra.purs

```diff
@@ -64,6 +68,30 @@ renderSubmodule ::
   ComponentHTML action slots m
 renderSubmodule lens wrapper renderer state = bimap (map wrapper) wrapper (renderer (view lens state))
 
+-- | This lets you map the state of a submodule that may not exist,
+-- | given an affine traversal into that optional substate. It's
+-- | an ugly solution, and it suffers from the same problem with
+-- | Halogen's `imapState` noted above. But for now, at least it works.
+mapMaybeSubmodule ::
+  forall m state state' action action' slots msg.
+  MonadAff m =>
+  state ->
+  Traversal' state state' ->
+  (action' -> action) ->
+  state' ->
+  HalogenM state' action' slots msg m Unit ->
+  HalogenM state action slots msg m Unit
+mapMaybeSubmodule state traversal wrapper submoduleDefaultState submoduleHandleAction = do
+  mSubmoduleState <- peruse traversal
+  for_ mSubmoduleState \submoduleState ->
+    (Halogen.imapState subToMain mainToSub <<< mapAction wrapper) $ submoduleHandleAction
+  where
+  subToMain submoduleState = set traversal submoduleState state
+
+  mainToSub mainState = case preview traversal mainState of
+    Just submoduleState -> submoduleState
+    _ -> submoduleDefaultState
```

Comment:
I think 91-93 could also be written:
```mainToSub mainState = fromMaybe submoduleDefaultState $ preview traversal mainState```
or even:
```mainToSub = fromMaybe submoduleDefaultState <<< preview traversal```
Not sure about parenthesis in last one

Reply:
So it can. Nice.